### PR TITLE
Partial velocities no longer require frame re-expression.

### LINKED ISSUE #10865: partial_velocity should not require a rexpression in the provided reference frame
I think that we may have a fundamental design flaw with some things about physics.vector. To illustrate, let's say that we have defined a vector that describes the linear velocity of point P in the reference frame N in terms of the generalized speeds: u1, u2, and u3.

```python
N_v_P = u1 * A.x + u2 * A.y + u3 * A.z
```

Note that we have this velocity vector is expressed in a frame A which has an arbitrary relationship with the frame N.

If point P happens to be an essential point to form the equations of motion with Kane's method, then we will need to take the partial velocity of the vector with respect to the generalized speeds in an intertial reference frame. If N happens to be that inertial reference frame, then we should be able to do:

```python
N_v_P.diff(u1, N)
```
and this should return:

```python
A.x
```
The partial velocity of N_v_P in N with respect to u1 is simply A.x. Of course, if A is oriented with respect to N in some way, then A.x can be expressed in N's unit vectors, but it isn't necessary to do that to compute the partial velocity.

The following code shows that this does not work:

```
moorepants@garuda:pydy(error-messages)$ ipython
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.1.2 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import sympy.physics.mechanics as me

In [2]: u1, u2, u3 = me.dynamicsymbols('u1, u2, u3')

In [3]: N = me.ReferenceFrame('N')

In [4]: A = me.ReferenceFrame('A')

In [5]: N_v_P = u1 * A.x + u2 * A.y + u3 * A.z

In [7]: N_v_P.diff(u1, N)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-7a11f17c841e> in <module>()
----> 1 N_v_P.diff(u1, N)

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/vector.py in diff(self, wrt, otherframe)
    519                 outvec += Vector([(v[0].diff(wrt), otherframe)])
    520             else:
--> 521                 if otherframe.dcm(v[1]).diff(wrt) == zeros(3, 3):
    522                     d = v[0].diff(wrt)
    523                     outvec += Vector([(d, v[1])])

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in dcm(self, otherframe)
    405         if otherframe in self._dcm_cache:
    406             return self._dcm_cache[otherframe]
--> 407         flist = self._dict_list(otherframe, 0)
    408         outdcm = eye(3)
    409         for i in range(len(flist) - 1):

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in _dict_list(self, other, num)
    250             return outlist[0]
    251         raise ValueError('No Connecting Path found between ' + self.name +
--> 252                          ' and ' + other.name)
    253 
    254     def _w_diff_dcm(self, otherframe):

ValueError: No Connecting Path found between N and A

In [9]: me.partial_velocity([N_v_P], [u1, u2, u3], N)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-56e053551919> in <module>()
----> 1 me.partial_velocity([N_v_P], [u1, u2, u3], N)

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/functions.py in partial_velocity(vel_list, u_list, frame)
    564         pvlist = []
    565         for j in u_list:
--> 566             vel = i.diff(j, frame)
    567             pvlist += [vel]
    568         list_of_pvlists += [pvlist]

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/vector.py in diff(self, wrt, otherframe)
    519                 outvec += Vector([(v[0].diff(wrt), otherframe)])
    520             else:
--> 521                 if otherframe.dcm(v[1]).diff(wrt) == zeros(3, 3):
    522                     d = v[0].diff(wrt)
    523                     outvec += Vector([(d, v[1])])

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in dcm(self, otherframe)
    405         if otherframe in self._dcm_cache:
    406             return self._dcm_cache[otherframe]
--> 407         flist = self._dict_list(otherframe, 0)
    408         outdcm = eye(3)
    409         for i in range(len(flist) - 1):

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in _dict_list(self, other, num)
    250             return outlist[0]
    251         raise ValueError('No Connecting Path found between ' + self.name +
--> 252                          ' and ' + other.name)
    253 
    254     def _w_diff_dcm(self, otherframe):

ValueError: No Connecting Path found between N and A
```

The problem is that the Vector object has no awareness of that it is a velocity of some point in some reference frame. Because of this, it doesn't really make sense to ask it to differentiate itself in the reference frame N.

But our vector differentiation in frames does work. It works because we *require* that there be a per-specified orientation between reference frames A and N. If this orientation is specified the correct answer is returned:

```
In [10]: q1, q2, q3 = me.dynamicsymbols('q1, q2, q3')

In [11]: A.orient(N, 'Body', (q1, q2, q3), 'XYZ')

In [12]: N_v_P.diff(u1, N)
Out[12]: A.x
```

This requires re-expression of the vector in the N frame so that the measure numbers in the N frame of N_v_P can be differentiated in that frame. But we know that re-expression is not necessary to get the correct result.

This comes up when you have ignorable coordinates. For example, if you want to study a satellite that is moving in inertial reference frame N, you can often ignore the 6 coordinates associated with the primary degrees of freedom up front. Currently, our software does not allow us to easily do this because of the above mentioned limitation.

Some thoughts:

- Maybe we need to define differentiation of the position/orientatoins, velocity, and accelerations at only on the Point and RefernceFrame objects because these have the full information about a given velocity vector, i.e. those objects know what frame the vector is with respect to. The vector by itself is not aware of what frame it is wrt.
- Maybe we need a "null" orientation option that lets you connect frames A and N but not specify the generalized coordinates and direction cosine matrix that relates the two. Then the vector.diff() function can deal with this properly.
- The fact that Vectors can differentiate themselves without know what frame they are defined wrt, seems mathematically incorrect.
- It is important to note that re-expression of a vector in another frame is different that knowing what frame a velocity vector is associated with.
- Do we need a VelocityVector object that knows more info like the frame the velocity is wrt or the associated points?
- I think Autolev properly tracks this with their syntax, for example in `N_v_P> = u1 * A1> + u2 * A2> + u3 * A3>` both the left and right hand sides have semantic meaning. This defines a velocity vector in N of P with that vector expression. All of that info must be stored in a data structure behind the scenes. They must have different types of vectors.

@sympy/mechanics What do you all think about this? I feel like we have an odd hole here that may take more than a hack to properly fix.

tl;dr Differentiation of vectors wrt to generalized speeds should not require that orientations be set between the frames that are associated with the unit vectors in the velocity vector expression and the frame that the velocity is defined it.

### PR DESCRIPTION
**Pull request by moorepants**
Fixes #10865

When computing a partial velocity (i.e. partial derivative of a velocity vector
with respect to a generalized speed), it is known a priori that the generalized
speed does not appear in the direction cosine matrices that relate any of the
unit vectors in the vector expression, so the partials can be computed without
the need to re-express the velocity vector in the desired reference frame.

To allow this to function a flag has been added to the Vector.diff() method
that specifies whether or not the user knows if the variable in question is or
isn't present in the direction cosine matrices. Secondly, the partial_velocity
function utilizes this flag to ensure that the partials properly compute even
if the orientation of the reference frames associated with the velocity vector
is not specified.

This change will allow users to more cleanly solve problems where the location
and orientation of the body with respect to the Newtonian frame is ignorable,
for example a simple satellite.

### CODE REVIEW

**Comment by moorepants** [comment]
I'm getting an unrelated doctest fail. Is master failing?

**Comment by moorepants** [comment]
@sympy/mechanics This needs a review. Thanks.

**Comment by moorepants** [comment]
@gilbertgede @oliverlee @hazelnusse What do you all think of this fix?

File: sympy/physics/vector/vector.py [hunk 1]
```diff
@@ -510,24 +516,39 @@ def diff(self, wrt, otherframe):
...
-        for i, v in enumerate(self.args):
-            if v[1] == otherframe:
-                outvec += Vector([(v[0].diff(wrt), otherframe)])
+
+        var = sympify(var)
+        _check_frame(frame)
+
+        partial = Vector(0)
+
+        for i, vector_component in enumerate(self.args):
```
Comment (oliverlee) [inline sympy/physics/vector/vector.py]: Doesn't look like this enumerate is being used.

**Comment by oliverlee** [comment]
seems fine

File: sympy/physics/vector/vector.py (hunk 1 above)
Comment (moorepants) [inline sympy/physics/vector/vector.py]: Oh yeah, that's a remnant from the previous implementation.

**Comment by moorepants** [comment]
Enumerate is removed. If tests pass I will merge. Thanks.

**Comment by moorepants** [comment]
Ok, I'm merging this.

**Pull request merged**